### ISSUE TITLE
Need previewOnKeyDown Tool event

### ISSUE THREAD

**Issue by NeilBurkholder** [SketchUp/api-issue-tracker#607]
In tools the onKeyDown event does not fire if there is a shortcut set to use that key. This goes for arrow keys, function keys, delete, etc. In fact this includes all the keys except Ctrl, Shift, Alt, and Win keys.

It would be nice to be able to receive these key presses in a tool even if they are used as a shortcut. This would allow custom delete behavior for example. Also there should be an option to return true, which would prevent the the shortcut from being triggered.

https://forums.sketchup.com/t/safe-hot-keys/150147/10?u=neil_burkholder
https://forums.sketchup.com/t/fr-allow-plugin-developers-to-override-delete-key-in-their-custom-tools/32848?u=neil_burkholder

**Comment by thomthom** [comment]
If anything I'd prefer to allow tools to intercept keys instead of adding a new event.

Regardless of implementation detail - I've run into similar situation myself with Vertex Tools that has it's own delete functions and accept a rich set of user input. (Like inputbox in the viewport, but currently can only accept numeric values as letters will trigger shortcuts.)

**Comment by NeilBurkholder** [comment]
I agree that the onKeyDown event could be used. For the alt key you can already return true (handled) to prevent the Sketchup menu from being activated.

An interesting note on the number keys is that, they can be set as shortcuts. If they are, the onKeyDown still fires (as well as the shortcut), but the VCB doesn't receive the input.

**Comment by NeilBurkholder** [comment]
Obviously for this to be useful the event would not only need to receive the key press, but there would also need to be a way to handle the event to prevent the shortcut from executing.

**Comment by Eneroth3** [comment]
If this functionality is added to the existing method, there is a risk there are tools where there by accident is a return value that would prevent shortcuts from working.

Maybe we should add a `excpectHonorkKeyReturnValue` or method, similar to the `AppObserver#expectsStartupModelNotifications` method?

**Comment by DanRathbun** [comment]
Another thought I have is for specific shortcut suppression ...

```ruby
def block_shortcut( vitrual_key, repeat, flags, view )
  if vitrual_key == VK_DELETE
    @state > 0 ? true : false
  else
    false
  end
end
```

... which would only be called if global **`#block_shortcuts?`** returned **`false`**.

Coders could use any kind of conditional like a `case` etc., to determine the boolean behavior.

**Comment by NeilBurkholder** [comment]
I pretty much agree with what Dan wrote. The reason I suggested a preview event is because I'm familiar with how that works in [Dot Net](https://docs.microsoft.com/en-us/dotnet/api/system.windows.uielement.onpreviewkeydown?view=net-5.0).

Several benefits to this approach.

1. You could receive the key press before the shortcut is invoked.
2. You would have the **option** to handle the event and prevent the keypress from being routed to the shortcut, or the KeyDow/KeyUp events.

Importantly this would allow you to receive the key press whether or not the shortcut was blocked.
By adding a new event existing functionality would not be impaired.

**Comment by DanRathbun** [comment]
> The reason I suggested a preview event is because I'm familiar with how that works in Dot Net.

FTR, the proposed `block_shortcut` callback method would indeed be a prerequisite step. 
Ie, it would get called by the SketchUp engine prior to calling any of:

* `onKeyDown`
* `onKeyUp`
* `onReturn`
* `onCancel`
* `onUserText
* any of the "on" mouse button callbacks

It **must** occur in a "pre" way or a tool author would not be able to control the allowing or suppressing of shortcuts when appropriate.

**Comment by NeilBurkholder** [comment]
Ok same thing then just a different event name proposed. ie: block_shortcut as opposed to onPreviewKeyDown.

**Comment by Fredosixx** [comment]
I am not too much in favor of blocking shortcuts from within interactive tools. After all, this is what shortcut are used for: to switch from one tool to another one, even if it's from a Ruby tool. If it is allowed, then it is likely that it will be used.

Instead I would propose 2 evolutions

1. **Have a special treatment for DEL key**. Delete should be disabled from within a Ruby interactive tool. It does not make sense that a user could delete the selection from within a custom tool, which usually has other use-cases for the Delete key. In most of my plugins, I have to disable the selection or intercept Delete and perform an undo.
2.  **Disable shortcuts after_at least one valid_ character is entered in the VCB** (usually a number). This would allow to enter inputs based on number with any suffix. Currently, we are limited to a few ones (S, R, M, and some non-letter sign like +, -, *, /, ', "). This would encourage the usage of VCB in custom tools. Funnily, if a letter is not assigned to a shortcut, it is accepted in the VCB. Native tools also do it for allowing entering units.

**Comment by thomthom** [comment]
I like those suggestions Fredo.

**Comment by Eneroth3** [comment]
I also like Fredo's ideas here.

Blocking shortcuts generally can be confusing to end users. You have expectations of certain keys taking you to certain tools. If that doesn't work, you could feel trapped in the tool (and not think about using the toolbar because the shortcuts are in your muscle memory). I think we should ask ourselves what keys we'd want to use and when.

The delete key is mentioned in the first post and we seem to all have experienced the need to react to it. A new `onDelete` method in the `Tool` interface could handle this. (And to get into details, I'd like the default behavior when this method is missing to be for delete to erase the selection. This is after all what Delete does in native tools like Rectangle, Scale, Circle etc so it's a reasonable fallback behavior).

Arrow keys were also brought up. These are in SketchUp already "owned" by the tool. Setting custom shortcuts to use the arrow keys would break key functionality (no pun intended) so here I think the solution is to disallow it being used as a shortcut in the first place.

Are there more keys we'd like to use, and in such case when?

Regarding the VCB I agree completely it needs cleaning up. Currently SketchUp has a of whitelist of allowed characters that SketchUp itself uses. The S suffix is for "sides" or "segments" and lets you change this property in Arc, Circle and polygon tool, even in a tool state where the VCB otherwise is for a different property. Y was added recently for Yards. M is for meters and R for radius I believe. rather than relying on a specific list (that isn't documented btw and in theory could change) I agree we should allow any characters here once the user has started typing a number.

**Comment by Eneroth3** [comment]
Allowing any character to the VCB logged as SKEXT-2918.

**Comment by NeilBurkholder** [comment]
A very good solution. That goes well with the current user interface. For example when a field is highlighted in one of the tray fields your don't expect spacebar to activate the select tool.

**Comment by DanRathbun** [comment]
One of the main points of the initial post is that the tool does not have (currently) a chance to react to shortcut keys or certain keys like DELETE. (But this is not specifically about the DELETE key.)

> rather than relying on a specific list (that isn't documented btw and in theory could change) I agree we should allow any characters here once the user has started typing a number.

Agree somewhat. Also if a minus, a plus etc. starts the character sequence.

How about introduce a directive entry? The user would start a directive with an exclamation (!) character, and then all shortcuts would be suppressed only for that entry.

**Comment by Eneroth3** [comment]
> One of the main points of the initial post is that the tool does not have (currently) a chance to react to shortcut keys or certain keys like DELETE. (But this is not specifically about the DELETE key.)

A general way to intercept and override shortcuts could allow for some unexpected behaviors for the end user. If misused by an extension developer who isn't a SketchUp user themselves, the end user could feel "trapped" in a tool.

I'm not saying that door is closed, but I think its worth exploring if it is really needed, or if we can cater for our needs with a special delete callback and more consistent VCB behavior. Feel free to add examples where intercepting shortcut would be needed.

**Comment by Eneroth3** [comment]
> Agree somewhat. Also if a minus, a plus etc. starts the character sequence.

The minus key can't be set as a shortcut, but is reserved for VCB input.

![2021-02-24_10h59_56](https://user-images.githubusercontent.com/5631725/108984005-bf7d6880-768f-11eb-8b0a-911e4614ac7d.png)

I don't know SketchUp currently treats +in shortcuts, as the shortcut interface doesn't honor my Swedish keyboard layout.

**Comment by Fredosixx** [comment]
The simplest is that the VCB accepts any character after a first valid one is entered (and thus appears in the VCB). 

We let the VCB apply its current logic for the first character. `!` and `?` seems to be accepted as a first character, as well as `+-*/`. So any character following them should be accepted too and not treated as shortcut.

Now, as Julia pointed out, there is a need to prevent a number of characters to be defined as shortcuts. Currently, you can assign short cuts to `=+-*!?` and even the `Del` key. This is indeed a potential issue, although in practice I believe no user would do that.

**Comment by DanRathbun** [comment]
Somehow this issue has been redirected into VCB behavior. This is not what the issue opener started the issue about.
It is also not specifically about the DELETE key or it's use as a shortcut. 
(These issues can have their **own** issue threads for tracking.)

> _A general way to intercept and override shortcuts could allow for some unexpected behaviors for the end user. If misused by an extension developer who isn't a SketchUp user themselves, the end user could feel "trapped" in a tool._

Well then there should always be an escape like CTRL+ESC (or F11) that always allows the user access to the exit.

**Comment by Eneroth3** [comment]
> Well then there should always be an escape like CTRL+ESC (or F11) that always allows the user access to the exit.

I don't think such behavior would make any difference. Users wouldn't know of this key combination, especially if it only makes a difference in a handful of oddly behaving extension.

Again, I'm curious if anyone can come up with a use case for intercepting shortcuts (other than delete and the separate VCB issue). I'm open for adding a more general API if there is a use case for it, but not very enthusiastic about having it for the sake of having it, especially giving the UX concerns.

**Comment by DanRathbun** [comment]
Can we change this issue thread to "Shortcuts with VCB use" or similar. It was very hard to find (and I knew it existed.)
